src/components/ble/weather/WeatherService.cpp

```diff
@@ -401,103 +409,52 @@ namespace Pinetime {
+          return reinterpret_cast<std::unique_ptr<T>&>(header);
         }
       }
 
-      return reinterpret_cast<std::unique_ptr<WeatherData::Clouds>&>(*this->nullHeader);
+      return reinterpret_cast<std::unique_ptr<T>&>(*this->nullHeader);
     }
 
-    std::unique_ptr<WeatherData::Obscuration>& WeatherService::GetCurrentObscuration() {
-      uint64_t currentTimestamp = GetCurrentUnixTimestamp();
-      for (auto&& header : this->timeline) {
-        if (header->eventType == WeatherData::eventtype::Obscuration && IsEventStillValid(header, currentTimestamp)) {
-          return reinterpret_cast<std::unique_ptr<WeatherData::Obscuration>&>(header);
-        }
-      }
+    std::unique_ptr<WeatherData::Clouds>& WeatherService::GetCurrentClouds() {
+      return GetEventOfType<WeatherData::Clouds>(WeatherData::eventtype::Clouds);
+    }
 
-      return reinterpret_cast<std::unique_ptr<WeatherData::Obscuration>&>(*this->nullHeader);
+    std::unique_ptr<WeatherData::Obscuration>& WeatherService::GetCurrentObscuration() {
+      return GetEventOfType<WeatherData::Obscuration>(WeatherData::eventtype::Obscuration);
     }
 
     std::unique_ptr<WeatherData::Precipitation>& WeatherService::GetCurrentPrecipitation() {
-      uint64_t currentTimestamp = GetCurrentUnixTimestamp();
-      for (auto&& header : this->timeline) {
-        if (header->eventType == WeatherData::eventtype::Precipitation && IsEventStillValid(header, currentTimestamp)) {
-          return reinterpret_cast<std::unique_ptr<WeatherData::Precipitation>&>(header);
-        }
-      }
-
-      return reinterpret_cast<std::unique_ptr<WeatherData::Precipitation>&>(*this->nullHeader);
+      return GetEventOfType<WeatherData::Precipitation>(WeatherData::eventtype::Precipitation);
     }
 
     std::unique_ptr<WeatherData::Wind>& WeatherService::GetCurrentWind() {
-      uint64_t currentTimestamp = GetCurrentUnixTimestamp();
-      for (auto&& header : this->timeline) {
-        if (header->eventType == WeatherData::eventtype::Wind && IsEventStillValid(header, currentTimestamp)) {
-          return reinterpret_cast<std::unique_ptr<WeatherData::Wind>&>(header);
-        }
-      }
-
-      return reinterpret_cast<std::unique_ptr<WeatherData::Wind>&>(*this->nullHeader);
+      return GetEventOfType<WeatherData::Wind>(WeatherData::eventtype::Wind);
     }
 
     std::unique_ptr<WeatherData::Temperature>& WeatherService::GetCurrentTemperature() {
-      uint64_t currentTimestamp = GetCurrentUnixTimestamp();
-      for (auto&& header : this->timeline) {
-        if (header->eventType == WeatherData::eventtype::Temperature && IsEventStillValid(header, currentTimestamp)) {
-          return reinterpret_cast<std::unique_ptr<WeatherData::Temperature>&>(header);
-        }
-      }
-
-      return reinterpret_cast<std::unique_ptr<WeatherData::Temperature>&>(*this->nullHeader);
+      return GetEventOfType<WeatherData::Temperature>(WeatherData::eventtype::Temperature);
```

Comment:
This might actually be slightly more involved than picking the first valid event.

Ideally we'd pick the most recently (ideally added, but that might be a waste) valid event, with the shortest expiricy but not the one soonest to expire. Goal being to pick the freshest forecast for the near immediate future rather than a daily forecast, should there be one.

Keeping the current behaviour with this PR would be fine though.

Reply:
Yeah, I did not want to change any behaviour, just doing some clean up.
Previously the invariant was that after insertion the events are tidied (removed expired and sorted) therefore the first pick was actually the nearest not yet expired one :^). My change broke this a wee bit, we still need to sort after insertion.